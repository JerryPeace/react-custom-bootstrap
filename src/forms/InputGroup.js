import PropTypes from 'prop-types';
import React from 'react';
import bootstrapUtils from '../utils/bootstrapUtils';
import classNames from 'classnames';
import TextField from './TextField';

class InputGroup extends React.Component {

    static propTypes = {
        bsClass: PropTypes.string,
        bsSize: PropTypes.oneOf(['lg', 'sm']),
        leftFieldType: PropTypes.oneOf(['button-group', 'others']),
        leftField: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        rightFieldType: PropTypes.oneOf(['button-group', 'others']),
        rightField: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        inputFieldRenderer: PropTypes.func
    };

    static defaultProps = {
        bsClass: 'input-group'
    };

    getValue() {
        return this.refs.inputCompo.getValue();
    }

    addonWrapper(component) {
        return (<span className='input-group-addon'>
                    {component}
                </span>);
    }

    btnGroupWrapper(component) {
        return (<span className='input-group-btn'>
                    {component}
                </span>);
    }

    renderSideComponent(field, fieldType) {

        if(field && fieldType) {

            if(fieldType === 'button-group') {
                return (this.btnGroupWrapper(field));
            } else {
                return (this.addonWrapper(field));
            }

        } else {
            return null;
        }
    }

    renderInputField(props) {
        const inputFieldRenderer = this.props.inputFieldRenderer;
        if (inputFieldRenderer) {
            return inputFieldRenderer(this.props);
        } else {
            return <TextField ref={'inputCompo'} {...props}/>;
        }
    }

    render() {
        const {className, leftField, leftFieldType, rightField, rightFieldType, bsClass, ...props} = this.props;

        let bsCls = bootstrapUtils.getCls(this.props);

        return (
            <div className={classNames(className, bsCls)}>
                {this.renderSideComponent(leftField, leftFieldType)}
                {this.renderInputField(props)}
                {this.renderSideComponent(rightField, rightFieldType)}
            </div>
        );
    }

}

export default InputGroup;
