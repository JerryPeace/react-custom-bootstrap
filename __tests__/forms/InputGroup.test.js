import React from 'react';
import { mount } from 'enzyme';
import InputGroup from 'forms/InputGroup';
import toJSON from'enzyme-to-json';

describe('InputGroup', () => {

    let inputGroup, inst;

    describe('InputGroup', () => {

        let getValue;

        beforeEach(() => {
            getValue = jest.fn();
            inputGroup = mount(<InputGroup />);
            inst = inputGroup.instance();
        });

        test('should render corrently', () => {
            expect(toJSON(inputGroup)).toMatchSnapshot();
        });

        test('get textfield value', () => {

            const input = inputGroup.find('input');
            input.simulate('focus');
            input.simulate('change', { target: { value: 'Hello' } });
            inputGroup.update();
            expect(input.node.value).toEqual(inst.getValue());

        });
    });

    describe('with only left field', () => {

        beforeEach(() => {
            inputGroup = mount(<InputGroup leftField='left field' leftFieldType='others' />);
        });

        test('should render corrently', () => {
            expect(toJSON(inputGroup)).toMatchSnapshot();
        });

    });

    describe('with only left field (button-group)', () => {

        beforeEach(() => {
            inputGroup = mount(<InputGroup leftField='left field' leftFieldType='button-group' />);
        });

        test('should render corrently', () => {
            expect(toJSON(inputGroup)).toMatchSnapshot();
        });

    });

    describe('with only right field', () => {

        beforeEach(() => {
            inputGroup = mount(<InputGroup rightField='right field' rightFieldType='others' />);
        });

        test('should render corrently', () => {
            expect(toJSON(inputGroup)).toMatchSnapshot();
        });

    });

    describe('with only right field (button-group)', () => {

        beforeEach(() => {
            inputGroup = mount(<InputGroup rightField='right field' rightFieldType='button-group' />);
        });

        test('should render corrently', () => {
            expect(toJSON(inputGroup)).toMatchSnapshot();
        });

    });

    describe('with both left field and right field', () => {

        beforeEach(() => {
            inputGroup = mount(<InputGroup
                leftField='left field' leftFieldType='others'
                rightField='right field' rightFieldType='others' />);
        });

        test('should render corrently', () => {
            expect(toJSON(inputGroup)).toMatchSnapshot();
        });

    });

});
