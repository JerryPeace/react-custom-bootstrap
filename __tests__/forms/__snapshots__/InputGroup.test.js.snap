// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroup InputGroup should render corrently 1`] = `
<InputGroup
  bsClass="input-group"
>
  <div
    className="input-group"
  >
    <Textfield
      bsClass="input"
      delayValidateTime={500}
      disabled={false}
      helpText=""
      helpTextClassName="has-error"
      inputType="text"
      labelAlign="left"
      labelClassName="col-xs-2"
      multiline={false}
      standalone={true}
      validator={[Function]}
      wrapperClassName="col-xs-10"
    >
      <LayoutHelper
        FieldBase={
          <input
            className=""
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        }
        bsClass="input"
        delayValidateTime={500}
        disabled={false}
        getInputEl={[Function]}
        helpText={null}
        helpTextClassName="has-error"
        inputType="text"
        labelAlign="left"
        labelClassName="col-xs-2"
        multiline={false}
        singleCheckboxRadio={false}
        standalone={true}
        validator={[Function]}
        wrapperClassName="col-xs-10"
      >
        <div
          className=""
        >
          <input
            className="form-control"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </div>
      </LayoutHelper>
    </Textfield>
  </div>
</InputGroup>
`;

exports[`InputGroup with both left field and right field should render corrently 1`] = `
<InputGroup
  bsClass="input-group"
  leftField="left field"
  leftFieldType="others"
  rightField="right field"
  rightFieldType="others"
>
  <div
    className="input-group"
  >
    <span
      className="input-group-addon"
    >
      left field
    </span>
    <Textfield
      bsClass="input"
      delayValidateTime={500}
      disabled={false}
      helpText=""
      helpTextClassName="has-error"
      inputType="text"
      labelAlign="left"
      labelClassName="col-xs-2"
      multiline={false}
      standalone={true}
      validator={[Function]}
      wrapperClassName="col-xs-10"
    >
      <LayoutHelper
        FieldBase={
          <input
            className=""
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        }
        bsClass="input"
        delayValidateTime={500}
        disabled={false}
        getInputEl={[Function]}
        helpText={null}
        helpTextClassName="has-error"
        inputType="text"
        labelAlign="left"
        labelClassName="col-xs-2"
        multiline={false}
        singleCheckboxRadio={false}
        standalone={true}
        validator={[Function]}
        wrapperClassName="col-xs-10"
      >
        <div
          className=""
        >
          <input
            className="form-control"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </div>
      </LayoutHelper>
    </Textfield>
    <span
      className="input-group-addon"
    >
      right field
    </span>
  </div>
</InputGroup>
`;

exports[`InputGroup with only left field (button-group) should render corrently 1`] = `
<InputGroup
  bsClass="input-group"
  leftField="left field"
  leftFieldType="button-group"
>
  <div
    className="input-group"
  >
    <span
      className="input-group-btn"
    >
      left field
    </span>
    <Textfield
      bsClass="input"
      delayValidateTime={500}
      disabled={false}
      helpText=""
      helpTextClassName="has-error"
      inputType="text"
      labelAlign="left"
      labelClassName="col-xs-2"
      multiline={false}
      standalone={true}
      validator={[Function]}
      wrapperClassName="col-xs-10"
    >
      <LayoutHelper
        FieldBase={
          <input
            className=""
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        }
        bsClass="input"
        delayValidateTime={500}
        disabled={false}
        getInputEl={[Function]}
        helpText={null}
        helpTextClassName="has-error"
        inputType="text"
        labelAlign="left"
        labelClassName="col-xs-2"
        multiline={false}
        singleCheckboxRadio={false}
        standalone={true}
        validator={[Function]}
        wrapperClassName="col-xs-10"
      >
        <div
          className=""
        >
          <input
            className="form-control"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </div>
      </LayoutHelper>
    </Textfield>
  </div>
</InputGroup>
`;

exports[`InputGroup with only left field should render corrently 1`] = `
<InputGroup
  bsClass="input-group"
  leftField="left field"
  leftFieldType="others"
>
  <div
    className="input-group"
  >
    <span
      className="input-group-addon"
    >
      left field
    </span>
    <Textfield
      bsClass="input"
      delayValidateTime={500}
      disabled={false}
      helpText=""
      helpTextClassName="has-error"
      inputType="text"
      labelAlign="left"
      labelClassName="col-xs-2"
      multiline={false}
      standalone={true}
      validator={[Function]}
      wrapperClassName="col-xs-10"
    >
      <LayoutHelper
        FieldBase={
          <input
            className=""
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        }
        bsClass="input"
        delayValidateTime={500}
        disabled={false}
        getInputEl={[Function]}
        helpText={null}
        helpTextClassName="has-error"
        inputType="text"
        labelAlign="left"
        labelClassName="col-xs-2"
        multiline={false}
        singleCheckboxRadio={false}
        standalone={true}
        validator={[Function]}
        wrapperClassName="col-xs-10"
      >
        <div
          className=""
        >
          <input
            className="form-control"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </div>
      </LayoutHelper>
    </Textfield>
  </div>
</InputGroup>
`;

exports[`InputGroup with only right field (button-group) should render corrently 1`] = `
<InputGroup
  bsClass="input-group"
  rightField="right field"
  rightFieldType="button-group"
>
  <div
    className="input-group"
  >
    <Textfield
      bsClass="input"
      delayValidateTime={500}
      disabled={false}
      helpText=""
      helpTextClassName="has-error"
      inputType="text"
      labelAlign="left"
      labelClassName="col-xs-2"
      multiline={false}
      standalone={true}
      validator={[Function]}
      wrapperClassName="col-xs-10"
    >
      <LayoutHelper
        FieldBase={
          <input
            className=""
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        }
        bsClass="input"
        delayValidateTime={500}
        disabled={false}
        getInputEl={[Function]}
        helpText={null}
        helpTextClassName="has-error"
        inputType="text"
        labelAlign="left"
        labelClassName="col-xs-2"
        multiline={false}
        singleCheckboxRadio={false}
        standalone={true}
        validator={[Function]}
        wrapperClassName="col-xs-10"
      >
        <div
          className=""
        >
          <input
            className="form-control"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </div>
      </LayoutHelper>
    </Textfield>
    <span
      className="input-group-btn"
    >
      right field
    </span>
  </div>
</InputGroup>
`;

exports[`InputGroup with only right field should render corrently 1`] = `
<InputGroup
  bsClass="input-group"
  rightField="right field"
  rightFieldType="others"
>
  <div
    className="input-group"
  >
    <Textfield
      bsClass="input"
      delayValidateTime={500}
      disabled={false}
      helpText=""
      helpTextClassName="has-error"
      inputType="text"
      labelAlign="left"
      labelClassName="col-xs-2"
      multiline={false}
      standalone={true}
      validator={[Function]}
      wrapperClassName="col-xs-10"
    >
      <LayoutHelper
        FieldBase={
          <input
            className=""
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        }
        bsClass="input"
        delayValidateTime={500}
        disabled={false}
        getInputEl={[Function]}
        helpText={null}
        helpTextClassName="has-error"
        inputType="text"
        labelAlign="left"
        labelClassName="col-xs-2"
        multiline={false}
        singleCheckboxRadio={false}
        standalone={true}
        validator={[Function]}
        wrapperClassName="col-xs-10"
      >
        <div
          className=""
        >
          <input
            className="form-control"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </div>
      </LayoutHelper>
    </Textfield>
    <span
      className="input-group-addon"
    >
      right field
    </span>
  </div>
</InputGroup>
`;
